lib/server/render.js

```diff
@@ -0,0 +1,132 @@
+const fs = require('fs-extra');
+const React = require('react');
+const renderToStaticMarkup = require('react-dom/server').renderToStaticMarkup;
+
+const path = require('path');
+const join = path.join;
+
+const env = require('./env.js');
+const readMetadata = require('./readMetadata.js');
+const toSlug = require('../core/toSlug.js');
+const DocsLayout = require('../core/DocsLayout.js');
+const MarkdownBlock = require('../core/MarkdownBlock.js');
+
+const CWD = process.cwd();
+const siteConfig = require(CWD + '/siteConfig.js');
+
+class Markdown {
+  // links to markdown files => links to their website htmls
+  mdToHtml(Metadata) {
+    const mdToHtml = {};
+    Object.keys(Metadata).forEach(id => {
+      const metadata = Metadata[id];
+      if (metadata.language !== 'en' || metadata.original_id) {
+        return;
+      }
+      let htmlLink =
+        siteConfig.baseUrl + metadata.permalink.replace('/next/', '/');
+      if (htmlLink.includes('/docs/en/')) {
+        htmlLink = htmlLink.replace('/docs/en/', '/docs/en/VERSION/');
+      } else {
+        htmlLink = htmlLink.replace('/docs/', '/docs/VERSION/');
+      }
+      mdToHtml[metadata.source] = htmlLink;
+    });
+    return mdToHtml;
+  }
+}
+
+class Docs {
+  // determine what file to use according to its id
+  metaToFile(metadata) {
+    const language = metadata.language;
+    if (metadata.original_id) {
+      return env.translation.enabled && metadata.language !== 'en'
+        ? join(CWD, 'translated_docs', metadata.language, metadata.source)
+        : join(CWD, 'versioned_docs', metadata.source);
+    } else {
+      return env.translation.enabled && metadata.language !== 'en'
+        ? join(CWD, 'translated_docs', metadata.language, metadata.source)
+        : join(CWD, '..', readMetadata.getDocsPath(), metadata.source);
+    }
+  }
+
+  tableOfContent(content) {
+    const TABLE_OF_CONTENTS_TOKEN = '<AUTOGENERATED_TABLE_OF_CONTENTS>';
+
+    const regexp = /\n###\s+(`.*`.*)\n/g;
+    let match;
+    const headers = [];
+    while ((match = regexp.exec(content))) {
+      headers.push(match[1]);
+    }
+
+    return headers
+      .map(header => `  - [${header}](#${toSlug(header)})`)
+      .join('\n');
+  }
+
+  render(metadata, mdToHtml) {
```

Comment:
I do like having this as a parameter. This means that we can do one calculation of the mdToHtml when coming from generate.js, however if requesting a single page, then each request will have to create it's own mdToHtml object anyway.

Maybe we can have this as an optional argument?

Reply:
Is there not a way that mdToHtml can be cached and imported as a require?

Or even better, can the information be stored in the normal metaData?

We are iterating over a hashTable storage because we are checking if the requested document contains a link to all other documents we have.

If we could match anything that looks like a link and specificially look that up, then we could instantly get the matching link rather than having to iterate through everything.

Reply:
Yes I agree there are more can be done to mdToHtml. Just thought to leave it for later until we feel confident about all the use cases.

Reply:
I'd probably consider seeing if we can get it on the React component level as it is to do with deciding what to display in the text.

But totally right, out of this scope.